app/src/main/java/org/kiwix/kiwixmobile/local_file_transfer/WifiDirectManager.kt

```diff
@@ -0,0 +1,424 @@
+      displayToast(string.severe_loss_error, Toast.LENGTH_LONG)
+    }
+  }
+
+  /* From WifiP2pManager.PeerListListener callback-interface */
+  override fun onPeersAvailable(peers: WifiP2pDeviceList) {
+    callbacks.updateListOfAvailablePeers(peers)
+  }
+
+  /* From WifiP2pManager.ConnectionInfoListener callback-interface */
+  override fun onConnectionInfoAvailable(groupInfo: WifiP2pInfo) { /* Devices have successfully connected, and 'info' holds information about the wifi p2p group formed */
+    this.groupInfo = groupInfo
+    performHandshakeWithSelectedPeerDevice()
+  }
+
+  val isGroupFormed: Boolean
+    get() = groupInfo!!.groupFormed
+
+  val isGroupOwner: Boolean
+    get() = groupInfo!!.isGroupOwner
+
+  val groupOwnerAddress: InetAddress
+    get() = groupInfo!!.groupOwnerAddress
+
+  fun sendToDevice(senderSelectedPeerDevice: WifiP2pDevice) { /* Connection can only be initiated by user of the sender device, & only when transfer has not been started */
+    if (!isFileSender || hasSenderStartedConnection) {
+      return
+    }
+    this.senderSelectedPeerDevice = senderSelectedPeerDevice
+    alertDialogShower.show(
+      FileTransferConfirmation(senderSelectedPeerDevice.deviceName), {
+        hasSenderStartedConnection = true
+        connect()
+        displayToast(string.performing_handshake, Toast.LENGTH_LONG)
+      })
+  }
+
+  private fun connect() {
+    if (senderSelectedPeerDevice == null) {
+      Log.d(TAG, "No device set as selected")
+    }
+    val config = WifiP2pConfig()
+    config.deviceAddress = senderSelectedPeerDevice!!.deviceAddress
+    config.wps.setup = WpsInfo.PBC
+    manager!!.connect(channel, config, object : ActionListener {
+      override fun onSuccess() { // UI updated from broadcast receiver
+      }
+
+      override fun onFailure(reason: Int) {
+        val errorMessage = getErrorMessage(reason)
+        Log.d(
+          TAG,
+          activity.getString(string.connection_failed) + ": " + errorMessage
+        )
+        displayToast(string.connection_failed, Toast.LENGTH_LONG)
+      }
+    })
+  }
+
+  private fun performHandshakeWithSelectedPeerDevice() {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Starting handshake")
+    }
+    peerGroupHandshakeAsyncTask = PeerGroupHandshakeAsyncTask(this)
+    peerGroupHandshakeAsyncTask!!.execute()
+  }
+
+  val totalFilesForTransfer: Int
+    get() = filesForTransfer!!.size
+
+  fun getFilesForTransfer(): ArrayList<FileItem> = filesForTransfer!!
+
+  fun setFilesForTransfer(fileItems: ArrayList<FileItem>) {
+    filesForTransfer = fileItems
+  }
+
+  val zimStorageRootPath: String
+    get() = sharedPreferenceUtil.prefStorage + "/Kiwix/"
+
+  fun getFileReceiverDeviceAddress(): InetAddress = fileReceiverDeviceAddress!!
+
+  fun setClientAddress(clientAddress: InetAddress?) {
+    if (clientAddress == null) { // null is returned only in case of a failed handshake
+      displayToast(string.device_not_cooperating, Toast.LENGTH_LONG)
```

Comment:
make sure all strings are referenced with `R`, I need to investigate why kotlin is sometimes suggesting importing the R but it is important for linting that the full thing be there in the source file eg `R.string.myString`. If it is `string.myString` lint can report it being unused